modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorage.java

```diff
@@ -944,15 +862,67 @@ public void finishCleanup() {
         }
     }
 
-    @Nullable VersionChain readVersionChain(RowId rowId) {
+    void throwExceptionIfStorageNotInRunnableState() {
+        StorageUtils.throwExceptionIfStorageNotInRunnableState(state.get(), this::createStorageInfo);
+    }
+
+    /**
+     * Searches version chain by row ID and converts the found version chain to the result if found.
+     *
+     * @param rowId Row ID.
+     * @param function Function for converting the version chain to a result, function is executed under the read lock of the page on which
+     *      the version chain is located. If the version chain is not found, then {@code null} will be passed to the function.
+     */
+    <T> @Nullable T findVersionChain(RowId rowId, Function<VersionChain, T> function) {
         try {
-            return versionChainTree.findOne(new VersionChainKey(rowId));
+            return versionChainTree.findOne(new VersionChainKey(rowId), new TreeRowClosure<>() {
+                @Override
+                public boolean apply(BplusTree<VersionChainKey, VersionChain> tree, BplusIo<VersionChainKey> io, long pageAddr, int idx) {
+                    return true;
+                }
+
+                @Override
+                public T map(VersionChain treeRow) {
+                    return function.apply(treeRow);
+                }
+            }, null);
         } catch (IgniteInternalCheckedException e) {
-            throw new StorageException("Error getting version chain: [rowId={}, {}]", e, rowId, createStorageInfo());
+            if (e.getCause() instanceof StorageException) {
+                throw (StorageException) e.getCause();
+            }
+
+            throw new StorageException("Row version lookup failed: [rowId={}, {}]", e, rowId, createStorageInfo());
         }
     }
 
-    void throwExceptionIfStorageNotInRunnableState() {
-        StorageUtils.throwExceptionIfStorageNotInRunnableState(state.get(), this::createStorageInfo);
+    /**
+     * Organizes external synchronization of update operations for the same version chain.
+     *
+     * <p>NOTE: When you try to execute in the closures on the pages of the tree, it leads to a deadlock.
```

Comment:
I don't get this note. What's its purpose?
Generally speaking, if you get locks in the wrong order, there's a chance of having a deadlock. That's a  common knowledge. I recommend documenting an overall locks model somewhere, instead of doing such notes in seemingly random places

Reply:
I will try to describe it in the documentation for the class.